### PR TITLE
Store gp_dbid and gp_contentid in conf files.

### PR DESCRIPTION
**Pull request by ashwinstar**
Store gp_dbid and gp_contentid in conf files.
    
    Currently, gp_dbid and gp_contentid is passed as command line
    arguments for starting QD and QE. Since, the values are stored in
    master's catalog table, to get the right values, must start the master
    first. Hence, hard-coded dbid=1 was used for starting the master in
    admin mode always. This worked fine till dbid was not used for
    anything on-disk. But given dbid is used for tablespace path in GPDB
    6, startting the instance with wrong dbid, means inviting recovery
    time failues, data corruption or data loss situations. Dbid=1 will go
    wrong after failover to standby master as it has dbid != 1. This
    commit hence eliminate the need of passing the gp_dbid and
    gp_contentid on command line, instead while creating the instance the
    values are stored in conf files for the instance.
    
    This also helps to avoid passing gp_dbid as argument to pg_rewind,
    which needs to start target instance in single user mode to complete
    recovery before performing rewind operation.
    
    Plus, this eases during development to just use pg_ctl start and not
    require to correctly pass these values.
    
     - gp_contentid is stored in postgresql.conf file.
     - gp_dbid is stored in internal.auto.conf.
     - Introduce internal.auto.conf file created during
       initdb. internal.auto.conf is included from postgresql.conf file.
     - Separate file is chosen to write gp_dbid for ease handling during
       pg_rewind and pg_basebackup, as can exclude copying this file from
       primary to mirror, instead of trying to edit the contents of the
       same after copy during these operations. gp_contentid remains same
       for primary and mirror hence having it in postgresql.conf file
       makes senes. If gp_contentid is also stored in this new file
       internal.auto.conf then pg_basebackup needs to be passed contentid
       as well to write to this file.

**Other commits mostly deal with fallout and validation from the above**
- gpinitsystem: add mirror to catalog first and then create them.
- pg_basebackup: write the gp_dbid after backup.
- gpexpand: update gp_dbid and gp_contentid post directory copy.
- pg_upgrade: retain all configuration files for segment.
- Validation for gp_dbid and gp_contentid between QD catalog and QE.
- Update tests to avoid passing gp_dbid and gp_contentid.

## Here are some reminders before you submit the pull request
- [X] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [X] Pass `make installcheck`

### CODE REVIEW

File: gpMgmt/bin/gpstart [hunk 1]
```diff
@@ -95,15 +93,12 @@ class GpStart:
...
-                    logger.warning("****************************************************************************")
-                    logger.warning("Master-only start requested in a configuration with a standby master.")
-                    logger.warning("This is advisable only under the direct supervision of Greenplum support. ")
-                    logger.warning("This mode of operation is not supported in a production environment and ")
-                    logger.warning("may lead to a split-brain condition and possible unrecoverable data loss.")
-                    logger.warning("****************************************************************************")
-                else:
-                    logger.info('Master-only start requested in configuration without a standby master.')
+                logger.warning("****************************************************************************")
+                logger.warning("If Master-only start requested in a configuration with a standby master.")
```
Comment (hlinnaka) [inline gpMgmt/bin/gpstart]: This message looks funny.

File: src/backend/fts/ftsmessagehandler.c [hunk 2]
```diff
@@ -417,6 +418,27 @@ HandleFtsWalRepPromote(void)
...
+
+	if (sscanf(query_string, FTS_MSG_FORMAT,
+			   message_type, &dbid, &contid) != 3)
+	{
+		ereport(ERROR,
+				(errmsg("received invalid FTS query: %s", query_string)));
+	}
+
+	if (dbid != GpIdentity.dbid)
+		ereport(PANIC,
```
Comment (hlinnaka) [inline src/backend/fts/ftsmessagehandler.c]: PANIC seems a bit harsh..

File: src/include/postmaster/fts.h [hunk 3]
```diff
@@ -25,6 +25,9 @@
 #define FTS_MSG_SYNCREP_OFF "SYNCREP_OFF"
 #define FTS_MSG_PROMOTE "PROMOTE"
 
+#define FTS_MSG_MAX_LEN 100
```
Comment (hlinnaka) [inline src/include/postmaster/fts.h]: This need a comment to explain what this is used for.

File: src/bin/pg_rewind/filemap.c [hunk 4]
```diff
@@ -108,6 +108,9 @@ process_source_file(const char *path, file_type_t type, size_t newsize,
 	if (strstr(path, "/" PG_TEMP_FILES_DIR "/") != NULL)
 		return;
 
+	if (strstr(path, "pg_log""/") != NULL)
```
Comment (hlinnaka) [inline src/bin/pg_rewind/filemap.c]: This seems unrelated to the rest of the patch.

File: gpMgmt/bin/gpstart (hunk 1 above)
Comment (ashwinstar) [reply gpMgmt/bin/gpstart]: Agree :-) was lazy in python and shell script. Will reword it properly.

File: src/backend/fts/ftsmessagehandler.c (hunk 2 above)
Comment (ashwinstar) [reply src/backend/fts/ftsmessagehandler.c]: Not sure what else we can do here. FATAL on FTS probes/promote will have effect of either failing over to master or not able to promote mirror, which seem equally harsh. Just WARNING, may go unnoticed and cause issues with tablespace.

PANIC for assert enabled builds and WARNING in production, or any other better suggestions, @hlinnaka ?

File: src/bin/pg_rewind/filemap.c (hunk 4 above)
Comment (ashwinstar) [reply src/bin/pg_rewind/filemap.c]: Yes, will open separate PR for this change to ignore `pg_log` directory copy for `pg_rewind`. Encountered it during development of this code and was super confusing for logs to show-up in wrong segment.

File: src/bin/pg_rewind/filemap.c [hunk 5]
```diff
@@ -182,7 +185,7 @@ process_source_file(const char *path, file_type_t type, size_t newsize,
 				 * An exception: PG_VERSIONs should be identical, but avoid
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply src/bin/pg_rewind/filemap.c]: Sorry, I'm guilty for that. "in toto" is really intended here. Meaning, "copy the whole file" (https://en.wiktionary.org/wiki/in_toto). That was perhaps too cute, but since it's like that in upstream, we should leave it alone in GPDB.

[the remark this answers is not in the document]
Comment (dgkimura) [reply src/bin/pg_rewind/filemap.c]: TIL ;-)

File: src/bin/initdb/initdb.c [hunk 6]
```diff
@@ -1329,7 +1329,7 @@ setup_config(void)
 	char		repltok[MAXPGPATH];
 	char		path[MAXPGPATH];
 	const char *default_timezone;
-	char	   *autoconflines[3];
+	char	   *autoconflines[2];
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/bin/initdb/initdb.c]: nope, not on purpose and is actually wrong. Thanks for catching!

File: src/bin/pg_basebackup/pg_basebackup.c [hunk 7]
```diff
@@ -2413,3 +2423,33 @@ main(int argc, char **argv)
 
 	return 0;
 }
+
+static void
+WriteInternalConfFile(void)
+{
+	char		filename[MAXPGPATH];
+	FILE	   *cf;
+	char line_to_write[100];
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/bin/pg_basebackup/pg_basebackup.c]: Currently, we are writing is very simple very short string. So, went with static allocation instead of fancy `PQExpBuffer` and dynamic allocation. In future if need spans out to construct fancy data to write to this file, can be converted to `PQExpBuffer`.

**Comment by ashwinstar** [comment]
On Thu, Jan 17, 2019 at 12:34 PM David Kimura <notifications@github.com>
wrote:

> *@dgkimura* commented on this pull request.
> Is the expectation that internal.auto.conf will contain anything specific
> to a server that cannot be generalized to its primary/mirror pair?


Yes, thinking I should add Header to this file describe the intent for the
file for future reference.

**Comment by ashwinstar** [comment]
On Thu, Jan 17, 2019 at 12:49 AM Heikki Linnakangas <
notifications@github.com> wrote:

> Are you going to squash the commits before pushing? They all seem like
> important parts of the overall change.


Yes, squashed the relevant commits for storing to config file which were
related pieces.
Kept validation separate as irrespective its independent unit and not
dependent on storing it in file. Similarly cleanup work and gpinitsystem
change for mirror registration is separate commit.

**Comment by BaiShaoqi** [comment]
great job👏
